### ISSUE TITLE
napi_module_register is not extern "C"

### ISSUE THREAD

**Issue by cosinusoidally** [nodejs/abi-stable-node#133]
I assume napi_module_register should be extern "C" in order to expose a pure C interface to modules? Looking at https://github.com/nodejs/abi-stable-node/blob/7c049ad24e108eddd7756a8b6cbfe684bfda00d4/src/node_jsvmapi.h#L46 napi_module_register does not currently appear to be extern "C"

I built the following dummy addon:
```
#include <node_jsvmapi.h>
#include <stdio.h>

void Init(napi_env env, napi_value exports, napi_value module,void *p) {
    printf("hello world\n");
    return;
}

NODE_MODULE_ABI(addon, Init)
```
I built it with the following command line (on an x86_64 Linux system with GCC 4.8.4):
```
g++ -g -std=gnu++0x -I. -fPIC -shared ./addon.cc  -o addon.node
```
(I had copied the headers node_jsvmapi.h, node_jsvmapi_internal.h, node_jsvmapi_types.h and node_macros.h in to the same dir as addon.cc).

Looking at the dynamic symbols:
```
root@localhost:~/node_headers# objdump -T addon.node

addon.node:     file format elf64-x86-64

DYNAMIC SYMBOL TABLE:
0000000000000000  w   D  *UND*  0000000000000000              __gmon_start__
0000000000000000  w   D  *UND*  0000000000000000              _Jv_RegisterClasses
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 puts
0000000000000000      D  *UND*  0000000000000000              _Z20napi_module_registerPv
0000000000000000  w   D  *UND*  0000000000000000              _ITM_deregisterTMCloneTable
0000000000000000  w   D  *UND*  0000000000000000              _ITM_registerTMCloneTable
0000000000000000  w   DF *UND*  0000000000000000  GLIBC_2.2.5 __cxa_finalize
00000000002010a8 g    D  .bss   0000000000000000  Base        _end
00000000002010a0 g    D  .data  0000000000000000  Base        _edata
00000000000007c5 g    DF .text  0000000000000027  Base        _Z4InitP10napi_env__P12napi_value__S2_Pv
00000000002010a0 g    D  .bss   0000000000000000  Base        __bss_start
0000000000000670 g    DF .init  0000000000000000  Base        _init
0000000000000800 g    DF .fini  0000000000000000  Base        _fini
```

We can see _Z20napi_module_registerPv (which is the C++ mangled name for napi_module_register). This gets called from _register_addon during initialisation (since the macro NODE_MODULE_ABI(addon, Init) expands into calling _register_addon with a napi_module_struct containing a pointer to Init). The  _register_addon function gets called during shared library load since it has `__attribute__((constructor))` set.

**Comment by jasongin** [comment]
Correct, `napi_module_register` should be extern "C".

I have [a PR](https://github.com/nodejs/abi-stable-node/pull/129/files) open where I'm refactoring some things related to `napi_module_register`. I'll update it to fix this bug also.

**Comment by jasongin** [comment]
https://github.com/nodejs/abi-stable-node/pull/129/commits/044ef5c6034350bdf367e564de37cb76d509bdf9

**Comment by digitalinfinity** [comment]
Fixed as of 14318a8 - closing
